objects/SetupChecker/script.lua

```diff
@@ -2471,41 +2434,46 @@ function updateDropdownSelection(id, value)
     end)
 end
 
-function addExpansionToggle(value)
-    return matchRecurse("expansionsRow", function (t)
-        table.insert(t.children[1].children[1].children, {
-            tag="Toggle",
-            value=value,
-            attributes={id = value, onValueChanged = "toggleExpansion"},
-            children={},
-        })
+-- Update a 2-width grid of toggle buttons to the have the given items and states
+-- @param id The id of the row containing the grid of toggles
+-- @param values A table mapping toggle names to boolean values
+-- @param onValueChanged The name of the function to call when a toggle is clicked
+function updateToggleGrid(id, values, onValueChanged)
+    return matchRecurse(id, function (t)
+        t.children[1].children[1].children = {}
+        for name,isOn in pairs(values) do
+            table.insert(t.children[1].children[1].children, {
+                tag="Toggle",
+                value=name,
+                attributes={id = name, onValueChanged = onValueChanged, isOn = tostring(isOn)},
+                children={},
+            })
+        end
         local count = #t.children[1].children[1].children
-        t.attributes["preferredHeight"] = math.floor((count + 1) / 2) * 60 + 0.1
+        t.attributes["preferredHeight"] = math.ceil(count / 2) * 60 + 0.1
     end)
 end
-function addEventToggle(value)
-    return matchRecurse("events", function (t)
-        table.insert(t.children[1].children[1].children, {
-            tag="Toggle",
-            value=value.." Events",
-            attributes={id = value.." Events", onValueChanged = "toggleEvents"},
-            children={},
-        })
-        local count = #t.children[1].children[1].children
-        t.attributes["preferredHeight"] = math.floor((count + 1) / 2) * 60 + 0.1
-    end)
+function updateExpansionToggles()
+    local exps = Global.getTable("expansions")
+    local values = {}
+    for name,_ in pairs(expansions) do
+        -- The Global expansions table stores nil for disabled expansions
+        -- We want a boolean false, so we explicitly check for equality to true
+        values[name] = (exps[name] == true)
+    end
+    return updateToggleGrid("expansionsRow", values, "toggleExpansion")
 end
-function removeToggle(id, value)
-    return matchRecurse(id, function (t)
-        for i,child in pairs(t.children[1].children[1].children) do
-            if child.value == value then
-                table.remove(t.children[1].children[1].children, i)
-                break
-            end
+function updateEventToggles()
+    local events = Global.getTable("events")
+    local values = {}
+    for name,guid in pairs(expansions) do
```

Comment:
could you not just iterate through the `events` table instead of checking each expansion for having events?

Reply:
As mentioned above, the `Global` `events` table maps an expansion to `nil` if it's disabled, and thus iterating over that table simply wouldn't bring up any expansion for which events are disabled.